In [108]:
#import libraries
import sqlite3
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from bs4 import BeautifulSoup

In [ ]:
#import table
conn = sqlite3.connect(r"F:\VScode L.3\The final project\City-Library-After-School-Program\library.db")

# tables overview
books = pd.read_sql_query("SELECT * FROM books ",conn)
checkouts = pd.read_sql_query("SELECT * FROM checkouts ",conn)
members = pd.read_sql_query("SELECT * FROM members ",conn)

print("BOOKS: \n",books)
print("------------------------------------")
print("CHECKOUT: \n",checkouts)
print("------------------------------------")
print("MEMBERS: \n",members)

BOOKS: 
     book_id                    title          author
0       501          The Silver Kite   Amina Darwish
1       502           Desert Compass   Amina Darwish
2       503        The Lantern Maker    Adel Roushdy
3       504      Rooftop Astronomers    Adel Roushdy
4       505      Letters to the Nile       Aya Hafez
5       506      The Paper Boat Club       Aya Hafez
6       507    Fossils and Fireflies     Dalia Serry
7       508    The Quiet Observatory     Dalia Serry
8       509      Marbles and Mirrors     Diaa Sultan
9       510    The Missing Metronome     Diaa Sultan
10      511     Winter in Alexandria    Farida Anwar
11      512   The Cartographer's Cat    Farida Anwar
12      513   Circuits for Beginners    Galal Mounir
13      514    The Robot in Row Nine    Galal Mounir
14      515       Songs of the Oasis      Hoda Bakry
15      516  The Beekeeper's Almanac      Hoda Bakry
16      517  Shadows on the Corniche     Hani Nagati
17      518     The Ninth Lighthouse 

In [ ]:
# SQL to DF

library_df = pd.read_sql_query("""
SELECT
    checkouts.checkout_id,
    checkouts.member_id,
    checkouts.book_id,
    checkouts.checkout_date,
    checkouts.return_date,
    members.first_name,
    members.last_name,
    members.grade,
    members.neighborhood,
    members.membership_status,
    members.join_date,
    books.title,
    books.author
FROM checkouts
JOIN members
    ON checkouts.member_id = members.member_id
JOIN books
    ON checkouts.book_id = books.book_id 
""", conn)

library_df.index.name = None
library_df.columns.name = None


print("LIBRARY merged_DB:\n", library_df )



library_df.to_csv("library_df.csv",index=False)



LIBRARY merged_DB:
      checkout_id  member_id  book_id checkout_date return_date first_name  \
0           9263       1047      517    2024-10-21  2024-11-07       Sara   
1           9340       1072      513    2025-08-24  2025-09-01       Seif   
2           9231       1053      523    2024-02-04  2024-02-16       Adam   
3           9129       1032      513    2025-06-21  2025-06-29       Nada   
4           9370       1079      511    2025-11-11  2025-12-03       Rana   
..           ...        ...      ...           ...         ...        ...   
386         9232       1044      513    2025-05-26  2025-06-11     Sherif   
387         9084       1008      511    2024-06-27  2024-07-22       Ziad   
388         9116       1024      519    2025-01-10  2025-02-09    Youssef   
389         9352       1076      501    2025-02-02         NaN       Dina   
390         9246       1044      529    2025-06-25  2025-07-04     Sherif   

    last_name  grade neighborhood membership_status   j

In [111]:
book_catalog= pd.read_json("books.json")
library_df= pd.read_csv(r"f:\VScode L.3\The final project\City-Library-After-School-Program\library_df.csv")

print(library_df["book_id"].dtype)
print(book_catalog["book_id"].dtype)

print(book_catalog["book_id"].duplicated().sum())

merged_detailed_data = library_df.merge (book_catalog, on="book_id",validate="many_to_one")

print("BEFORE:", len(library_df))
print("AFTER:", len(merged_detailed_data))

merged_detailed_data.to_csv("Merged (sql-json).csv",index=False)

int64
int64
0
BEFORE: 391
AFTER: 391


In [112]:
with open(r"F:\VScode L.3\The final project\City-Library-After-School-Program\summer_checkouts.html","r",encoding="utf-8") as file:
    summer_checkouts = file.read()

soup = BeautifulSoup(summer_checkouts,"html.parser")
print(soup.find_all("table"))

summer_checkouts = pd.read_html(r"F:\VScode L.3\The final project\City-Library-After-School-Program\summer_checkouts.html")
summer_checkouts = summer_checkouts[0]

summer_checkouts.columns = (
    summer_checkouts.columns
    .str.lower()
    .str.replace(" ", "_")
)


summer_checkouts.to_csv("scraped_data HTML.csv", index=False)

summer_checkouts_df = pd.read_csv(r"F:\VScode L.3\The final project\City-Library-After-School-Program\scraped_data HTML.csv")



print(summer_checkouts_df)

[<table>
<tr><th>Member ID</th><th>Book ID</th><th>Checkout Date</th></tr>
<tr><td>1026</td><td>522</td><td>2025-07-11</td></tr>
<tr><td>1049</td><td>520</td><td>2025-07-11</td></tr>
<tr><td>1062</td><td>525</td><td>2025-07-05</td></tr>
<tr><td>1065</td><td>520</td><td>2025-07-07</td></tr>
<tr><td>1104</td><td>515</td><td>2025-07-07</td></tr>
<tr><td>1009</td><td>503</td><td>2025-07-09</td></tr>
<tr><td>1063</td><td>522</td><td>2025-07-07</td></tr>
<tr><td>1022</td><td>511</td><td>2025-07-12</td></tr>
<tr><td>1029</td><td>523</td><td>2025-07-09</td></tr>
<tr><td>1201</td><td>509</td><td>2025-07-10</td></tr>
<tr><td>1005</td><td>513</td><td>2025-07-10</td></tr>
<tr><td>1104</td><td>526</td><td>2025-07-05</td></tr>
<tr><td>1058</td><td>518</td><td>2025-07-08</td></tr>
<tr><td>1002</td><td>521</td><td>2025-07-08</td></tr>
<tr><td>1150</td><td>530</td><td>2025-07-10</td></tr>
<tr><td>1041</td><td>504</td><td>2025-07-05</td></tr>
<tr><td>1055</td><td>526</td><td>2025-07-11</td></tr>
<tr><td

In [ ]:
summer_checkouts_df = pd.read_csv(r"F:\VScode L.3\The final project\City-Library-After-School-Program\scraped_data HTML.csv")
merged_detailed_data = pd.read_csv(r"F:\VScode L.3\The final project\City-Library-After-School-Program\Merged (sql-json).csv")

print(merged_detailed_data.columns.tolist())
print(summer_checkouts_df.columns.tolist())

final_df = merged_detailed_data.merge(
    summer_checkouts_df,
    on=["member_id", "book_id", "checkout_date"],
    how="left",
    validate="many_to_one"
)

final_df.to_csv("task1_combined_data.csv", index=False)

['checkout_id', 'member_id', 'book_id', 'checkout_date', 'return_date', 'first_name', 'last_name', 'grade', 'neighborhood', 'membership_status', 'join_date', 'title', 'author', 'genre', 'pages', 'publication_year', 'publisher']
['member_id', 'book_id', 'checkout_date']
['member_id', 'book_id', 'checkout_date']


In [ ]:
#import prepared data
df = pd.read_csv(r"F:\VScode L.3\The final project\City-Library-After-School-Program\task1_combined_data.csv")

In [117]:
#Explore the Data
print("SHAPE : \n",df.shape)
print("------------------------------------")
print("SIZE : \n",df.size)
print("------------------------------------")
print("COLUMNS NAMES : \n",df.columns)
print("------------------------------------")
print("HEAD : \n",df.head())
print("------------------------------------")
print("INFO : ")
df.info()
print("------------------------------------")
print("DESCRIBE : \n",df.describe())
print("------------------------------------")
print("DATA TYPES : \n",df.dtypes)

SHAPE : 
 (391, 17)
------------------------------------
SIZE : 
 6647
------------------------------------
COLUMNS NAMES : 
 Index(['checkout_id', 'member_id', 'book_id', 'checkout_date', 'return_date',
       'first_name', 'last_name', 'grade', 'neighborhood', 'membership_status',
       'join_date', 'title', 'author', 'genre', 'pages', 'publication_year',
       'publisher'],
      dtype='str')
------------------------------------
HEAD : 
    checkout_id  member_id  book_id checkout_date return_date first_name  \
0         9263       1047      517    2024-10-21  2024-11-07       Sara   
1         9340       1072      513    2025-08-24  2025-09-01       Seif   
2         9231       1053      523    2024-02-04  2024-02-16       Adam   
3         9129       1032      513    2025-06-21  2025-06-29       Nada   
4         9370       1079      511    2025-11-11  2025-12-03       Rana   

  last_name  grade neighborhood membership_status   join_date  \
0    Rashad    NaN   Heliopolis      

In [127]:
#Inspect the Data
##Problem 1: Missing Values
print("MISSING_VALUES_COUNT : \n",df.isna().sum())
print("------------------------------------")
##Problem 2: Duplicates That Aren't All the Same
print("CHECK FOR DUPLICATES : \n",df[df.duplicated(keep=False)])
print("DUPLICATED_VALUES_COUNT : ",df.duplicated().sum())
print("------------------------------------")
##Possible Problem 1: OUTLIERS
print("SKEW : \n", df.skew(numeric_only=True))
print("------------------------------------")
#================================================================
##Problem 3: The Same Value Written Different Ways
print("NEIGHBORHOODS : ",df['neighborhood'].unique())
print("------------------------------------")
print("BOOKS CATEGORIES : ",df['genre'].unique())
print("------------------------------------")
print("GRADES : ",df['grade'].unique())
print("------------------------------------")
print("MEMBERSHIP STATUSES : ",df['membership_status'].unique())
print("------------------------------------")
#================================================================
##Problem 4: Checkouts With No Matching Member
invalid_member_ids = ~df["member_id"].isin(members["member_id"])
print("INVALID MEMBER IDs : \n", df.loc[invalid_member_ids, "member_id"])
print("INVALID MEMBER IDs COUNT : ", invalid_member_ids.sum())
print("------------------------------------")
##Possible Problem 2: NEGATIVE VALUES
numeric_columns = df.select_dtypes(include="number").columns
print("NEGATIVE VALUES : \n",(df[numeric_columns] < 0).sum())

MISSING_VALUES_COUNT : 
 checkout_id           0
member_id             0
book_id               0
checkout_date         0
return_date          65
first_name            0
last_name             0
grade                36
neighborhood          0
membership_status     0
join_date             5
title                 0
author                0
genre                 0
pages                 0
publication_year     34
publisher             0
dtype: int64
------------------------------------
CHECK FOR DUPLICATES : 
      checkout_id  member_id  book_id checkout_date return_date first_name  \
69          9296       1065      501    2024-12-15  2025-01-11       Adam   
73          9280       1054      517    2024-09-06  2024-09-15      Retaj   
113         9052       1019      501    2024-06-20  2024-06-29    Mostafa   
116         9194       1024      513    2024-09-17  2024-10-09    Youssef   
121         9193       1034      519    2024-09-15  2024-10-01        Aya   
225         9296       1065   

### **PROBLEMS FOUND :**

***1. Missing values :***

  * `return_date` >> **65** missing values  
  * `grade` >> **36** missing values  
  * `join_date` >> **5** missing values  
  * `publication_year` >> **34** missing values  


***2. Duplicated rows :***

   * **8 duplicated rows** were found.  
    * The duplicated records are exact duplicates of existing rows.  
    * Duplicate row indices: **69, 73, 113, 116, 121, 225, 244, 246, 278, 301, 330, 333, 357, 375, 378, 390**  


***3. Text issues :***

   * `neighborhood` >> inconsistent capitalization and extra whitespace:
     * `Heliopolis` / `HELIOPOLIS`
     * `Zamalek` / `zamalek`
     * `Nasr City` / `NASR CITY`
     * `Maadi` / `Maadi `

   * `membership_status` >> inconsistent capitalization:
     * `Active` / `active`
     * `Inactive` / `inactive`


***4. Invalid member IDs :***

   * **0 invalid member IDs** were found.
     * All `member_id` values in the combined dataset match registered members.


***5. Negative values :***

   * No negative values were found in the numeric columns checked: